### PR TITLE
Raising PR to get code review for the backend code of VideoLibrary

### PR DESCRIPTION
**Pull request by sruthiragupathy**

### CODE REVIEW

File: Controllers/note.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
+const _ = require('lodash');
+const Note = require('../Models/note');
+
+const getAllNotesOfAUser = async (req, res) => {
+  const { userId } = req.params;
+  try {
+    const response = await Note.find({ userId });
+    res.json({ response });
+  } catch (error) {
+    res.status(400).json({ response: error.message });
```
Comment (poojasabnani) [inline Controllers/note.js:10]: consistency with API responses can be followed.

in category.js the response is different and in this file the response format is different. (status code missing in category.js APIs as well)

File: Controllers/param.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  try {
+    const user = await User.findById(id);
+    if (!user) {
+      throw Error('No such user found');
+    }
+    // console.log({ user });
+    req.user = user;
+    next();
+  } catch (error) {
+    return res.status(400).json({ success: true, error: error.message });
```
Comment (poojasabnani) [inline Controllers/param.js:14]: why is success status true if statusCode is 400?

File: Controllers/param.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+exports.getPlaylistByUser = async (req, res, next, id) => {
+  try {
+    const playlists = await Playlist.find({ userId: id });
+    // console.log({ playlists });
+    req.playlists = playlists;
+    next();
+  } catch (error) {
+    return res.status(400).json({ success: true, error: error.message });
```
Comment (poojasabnani) [inline Controllers/param.js]: same here..

File: Controllers/param.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+exports.getVideoById = async (req, res, next, id) => {
+  try {
+    const video = await Video.findById(id);
+    // console.log({ video });
+    req.video = video;
+    next();
+  } catch (error) {
+    return res.status(400).json({ success: true, error: error.message });
```
Comment (poojasabnani) [inline Controllers/param.js]: here as well

File: Controllers/playlist.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
+const mongoose = require('mongoose');
+const Playlist = require('../Models/playlist');
+// const Video = require('../Models/video');
```
Comment (poojasabnani) [inline Controllers/playlist.js:3]: can be cleaned up

File: Controllers/playlist.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    userId,
+    name,
+    videos: [{ _id: video._id, video: video._id }],
+  });
+  try {
+    let response = await newVideo.save();
+    response = await response.populate('videos.video').execPopulate();
+    res.json({ success: true, response });
+  } catch (error) {
+    res.json({ success: false, response: error.message });
```
Comment (poojasabnani) [inline Controllers/playlist.js]: error code can be added

File: Models/note.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const childSchema = new Schema({
+  note: {
+    type: String,
+    trim: true,
+    // required: 'note is required',
```
Comment (poojasabnani) [inline Models/note.js]: can be cleaned up

File: Routes/playlist.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  getPlaylistById,
+} = require('../Controllers/param');
+const {
+  getPlaylistsOfAUser,
+  createPlaylist,
+  addVideoToPlaylist,
+  removeVideoFromPlaylist,
+  removePlaylist,
+  getVideosOfAPlaylist,
+} = require('../Controllers/playlist');
```
Comment (poojasabnani) [inline Routes/playlist.js:15]: This is very well structured!

File: Routes/playlist.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+router.param('userId', getPlaylistByUser);
+router.param('videoId', getVideoById);
+router.param('playlistId', getPlaylistById);
+
+router.get('/:userId/playlists', getPlaylistsOfAUser);
+router.get('/playlist/:playlistId', getVideosOfAPlaylist);
+router.post('/:userId/playlists/:videoId', createPlaylist);
+router.post('/playlist/:playlistId/:videoId', addVideoToPlaylist);
+router.delete('/playlist/:playlistId/:videoId', removeVideoFromPlaylist);
+router.delete('/playlists/:playlistId', removePlaylist);
```
Comment (poojasabnani) [inline Routes/playlist.js:29]: nicely structured! very clean!

**Review by poojasabnani — COMMENTED** [review]
Hi Sruthi, 
Very well structured code and great naming convention.

Some things that can be addressed are:

1. Maintaining consistency with the response objects.
2. Removing the commented out code
3. Removing console.logs which were used for debugging.
4. Good to have : console.error can be used for errors

File: Controllers/category.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+const Category = require('../Models/category');
+
+exports.addCategory = async (req, res) => {
```
Comment (AnkitaBagale) [inline Controllers/category.js:3]: just a suggestion we can export multiple functions with module.exports = {Category, getCategories}

File: Controllers/note.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    try {
+      await notesOfAVideo.notes.push({
+        note,
+        time,
+      });
+      // console.log({notesOfAVideo});
+      await notesOfAVideo.save();
+      const response = await Note.find({ userId });
+      // console.log(response);
+      res.json({ response });
```
Comment (AnkitaBagale) [inline Controllers/note.js:28]: Please remove commented code

File: Controllers/playlist.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+  }
+};
+
+const addVideoToPlaylist = async (req, res) => {
+  const { playlist, video } = req;
+  // const { videoId } = req.params;
+  try {
+    if (!playlist.videos.id(video._id)) {
```
Comment (AnkitaBagale) [inline Controllers/playlist.js:71]: Please remove commented code

File: Models/Database.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
+exports.videoList = [
+  {
+    kind: 'youtube#searchResult',
```
Comment (AnkitaBagale) [inline Models/Database.js:3]: Please remove videoList if not in use

File: Models/video.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
+// const { ObjectID } = require('bson');
+const mongoose = require('mongoose');
```
Comment (AnkitaBagale) [inline Models/video.js:1]: Please remove console.log

File: Routes/note.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  updateNote,
+  deleteNote,
+} = require('../Controllers/note');
+
+const router = express.Router();
+
+router.get('/:userId/notes', getAllNotesOfAUser);
+router.post('/:userId/notes/:videoId', createNotesForAVideo);
+// router.put('/:userId/notes/:videoId', addNoteToExistingVideo);
+router.post('/:userId/notes/:videoId/:noteId', updateNote);
```
Comment (AnkitaBagale) [inline Routes/note.js:13]: Please remove commented code

File: Routes/video.js [hunk 16]
```diff
@@ -0,0 +1,9 @@
+const express = require('express');
+// const { addCategory } = require('../Controllers/category');
+const { getAllVideos, getVideoById } = require('../Controllers/video');
+
```
Comment (AnkitaBagale) [inline Routes/video.js:3]: Please remove commented code

File: normalizeData.js [hunk 17]
```diff
@@ -0,0 +1,57 @@
+const { default: axios } = require('axios');
```
Comment (AnkitaBagale) [inline normalizeData.js:1]: Is this file in use? If not we can add it to .gitignore

**Comment by AnkitaBagale** [comment]
Hi Sruthi,

Very well structured code and great naming convention.
Just one suggestion always remove consoles and comments from the code before committing to the repo.

File: Controllers/note.js (hunk 1 above)
Comment (sruthiragupathy) [reply Controllers/note.js:10]: This has been taken care of

File: Controllers/param.js (hunk 2 above)
Comment (sruthiragupathy) [reply Controllers/param.js:14]: done

File: Controllers/param.js (hunk 3 above)
Comment (sruthiragupathy) [reply Controllers/param.js]: done

File: Controllers/param.js (hunk 4 above)
Comment (sruthiragupathy) [reply Controllers/param.js]: done

File: Models/note.js (hunk 7 above)
Comment (sruthiragupathy) [reply Models/note.js]: done

**Comment by sruthiragupathy** [comment]
> Hi Sruthi,
> 
> Very well structured code and great naming convention.
> Just one suggestion always remove consoles and comments from the code before committing to the repo.

Thank you Ankita, Your points have been taken care of

**Pull request merged**